### PR TITLE
gguf: Add types for LLM architectures

### LINKED ISSUE #566: [gguf types] Add missing types & make existing types stronger
https://github.com/huggingface/huggingface.js/pull/562 adds initial typing to @huggingface/gguf package.

However, not all architectures are added in #562. (see [here](https://github.com/ggerganov/ggml/blob/master/docs/gguf.md#required) & [here](https://github.com/huggingface/huggingface.js/pull/562#discussion_r1531981246)). For example, one can add [mamba](https://github.com/ggerganov/ggml/blob/master/docs/gguf.md#mamba) & type mamba specific fields (i.e. `mamba.ssm.inner_size`. See RWKV-specific typing for inspiration https://github.com/huggingface/huggingface.js/blob/81e6ce19d119978ce827c3f166d601c8f24fe454/packages/gguf/src/types.ts#L117

### PR DESCRIPTION
**Pull request by ngxson**
Resolve #566

This PR introduces a generator script that pulls `llama.cpp` file, then use regex to extract needed information.

It also addresses some smaller issues:
- Some places (for example `Attention`, `Rope`, `MOE`,...) use type intersection while it should be union
- `TokenizerModel` should be limited to 4 models, see `llm_load_vocab()`: https://github.com/ggerganov/llama.cpp/blob/928e0b7013c862cf10701957b3d654aa70f11bd8/llama.cpp#L4198

While it's fun to have the script & these type definitions, I think it would be nice to have some clearer directions on how to use them in the future.

Demo:

image

### CODE REVIEW

File: packages/gguf/scripts/generate-llm.ts
Comment (julien-c) [inline packages/gguf/scripts/generate-llm.ts:8]: i was basing myself on the `gguf-py/gguf/constants.py` file (i.e. from the Python parser), but I guess the cpp code is at least as good a source of truth?

**Review by julien-c — COMMENTED** [review]
very cool work @ngxson

I've launched the CI run, let me know if any help is needed, LGTM otherwise

File: packages/gguf/scripts/generate-llm.ts
Comment (julien-c) [inline packages/gguf/scripts/generate-llm.ts:67]:
```suggestion
	const matchedArchList = cppSource.match(/LLM_ARCH_NAMES = ([^;]+)/)?.[1].split("\n") ?? [];
```

Comment (julien-c) [reply packages/gguf/scripts/generate-llm.ts:67]: (and similar linting nits)

**Comment by mishig25** [comment]
@ngxson locally you can check the errors since the CI uses same commands:
```
pnpm format
pnpm lint
```

File: packages/gguf/package.json
Comment (mishig25) [inline packages/gguf/package.json:34]:
```suggestion
		"build": "pnpm build:llm && tsup src/index.ts --format cjs,esm --clean --dts",
```

`build:llm` should be called here before build right? (cc: @coyotte508)

File: packages/gguf/src/transformer-llm.ts
Comment (mishig25) [inline packages/gguf/src/transformer-llm.ts:1]: since it is a generated file, should we add it into the .gitignore ?

File: packages/gguf/scripts/generate-llm.ts
Comment (mishig25) [inline packages/gguf/scripts/generate-llm.ts:69]:
```suggestion
		const matched = line.match(/(LLM_ARCH_[A-Z0-9_]+),\s+"(.+?)"/);
```

I think you can use the [greedy/lazy regex quantifiers](https://javascript.info/regexp-greedy-and-lazy). Let me know if this change works. If so, uou can replace all instances of `"([^"]+)"` with `"(.+?)"`

File: packages/gguf/scripts/generate-llm.ts
Comment (mishig25) [inline packages/gguf/scripts/generate-llm.ts:126]: `LLM_ARCH_[A-Z0-9_]+` part is also used on line 100 https://github.com/ngxson/huggingface.js/blob/5313453c75b3f72fd1381762aa6450d33f77749e/packages/gguf/scripts/generate-llm.ts#L100

therefore, we should declare this regex as const `cont RE_XYZ = /LLM_ARCH_[A-Z0-9_]+)/` & re-use this re in this line like `new Regexp(/case ${RE_XYZ}/)`

File: packages/gguf/scripts/generate-llm.ts
Comment (mishig25) [inline packages/gguf/scripts/generate-llm.ts:73]: to make the regexes even more readable, you can use [named capturing group](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Regular_expressions/Named_capturing_group) & use like `matched.groups.name rather than matched[2]`

File: packages/gguf/package.json
Comment (julien-c) [reply packages/gguf/package.json:34]: or maybe run it once a week from a GitHub Action or something?

Comment (coyotte508) [reply packages/gguf/package.json:34]: It fetches a remote file so yes - calling it with a cron job is better than at every install

Comment (ngxson) [reply packages/gguf/package.json:34]: I prefer to have it run manually, because this way we can verify the generated script before pushing. But if we can run a Github Action workflow and create new PR if there're changes, it's also a good idea.

File: packages/gguf/scripts/generate-llm.ts
Comment (ngxson) [reply packages/gguf/scripts/generate-llm.ts:8]: I'm not very familiar with `gguf-py`, but seem like only in cpp code that you can find the list of metadata used by model (in python library, there's only the list of tensors, not list of kv - those can be found in `convert*.py` scripts)

File: packages/gguf/scripts/generate-llm.ts
Comment (ngxson) [reply packages/gguf/scripts/generate-llm.ts:67]: Actually I used these unsafe casting because I'd rather to have the script failed, instead of having it generates bad output. But I'm OK with this proposal, I should have added checks and throw errors. I'll fix this :+1:

File: packages/gguf/scripts/generate-llm.ts
Comment (julien-c) [reply packages/gguf/scripts/generate-llm.ts:8]: yes that makes sense!

File: packages/gguf/scripts/generate-llm.ts
Comment (ngxson) [reply packages/gguf/scripts/generate-llm.ts:187]: This function is only used by the generator script, which should not be part of the distributed (transpiled) source code. Do you think it's still useful to move it into utils?

File: packages/gguf/scripts/generate-llm.ts
Comment (ngxson) [reply packages/gguf/scripts/generate-llm.ts]: Thanks for pointing me in the correct direction. I replaced all instances of `"([^"]+)"` with `"(.+?)"`. However, the same technique does not work with `([^;]+)` ==> `(.+?);` (matches `ABC;` for example). Please let me know if you have any idea. Thanks.

File: packages/gguf/scripts/generate-llm.ts
Comment (mishig25) [reply packages/gguf/scripts/generate-llm.ts:187]: good for now

**Comment by ngxson** [comment]
@julien-c @mishig25 Regarding how to actually use these types, I'm thinking about having a list of functions like this:

```ts
getArchLlama(metadata: GGUFMetadata): ArchLlama?
getArchFalcon(metadata: GGUFMetadata): ArchFalcon?
...
```

Which returns `undefined` if the model is not the one being checked.

The idea is to prevent user from doing something like `const arch = model as ArchLlama`, but I'm not sure if my proposal is the right way to do.

Do you have any other ideas? Thanks.

File: packages/gguf/scripts/generate-llm.ts
Comment (mishig25) [inline packages/gguf/scripts/generate-llm.ts:160]: why not put this comment line on top of the file?

File: packages/gguf/scripts/generate-llm.ts
Comment (mishig25) [inline packages/gguf/scripts/generate-llm.ts:42]: is this todo being handled in this PR or follow-up PR ?

File: packages/gguf/scripts/generate-llm.ts
Comment (mishig25) [inline packages/gguf/scripts/generate-llm.ts:69]:
```suggestion
	const matchedArchList = cppSource.match(/LLM_ARCH_NAMES = (?<names>[^;]+)/)?.groups?.names.split("\n");
	if (!matchedArchList?.length) {
```

File: packages/gguf/scripts/generate-llm.ts
Comment (ngxson) [reply packages/gguf/scripts/generate-llm.ts:160]: I usually place import on the top then the comment. Just personal preference though, but you're right, it should be on top of the file.

File: packages/gguf/scripts/generate-llm.ts
Comment (mishig25) [inline packages/gguf/scripts/generate-llm.ts:78]:
```suggestion
				tsName: snakeToPascal(matched.groups?.cppConst.replace("LLM_", "")),
```

File: packages/gguf/scripts/generate-llm.ts
Comment (mishig25) [inline packages/gguf/scripts/generate-llm.ts:96]: as in other regexs, let's use named groups here as well

File: packages/gguf/scripts/generate-llm.ts
Comment (mishig25) [inline packages/gguf/scripts/generate-llm.ts:104]: similar change is needed as in https://github.com/huggingface/huggingface.js/pull/640#discussion_r1583785883

File: packages/gguf/scripts/generate-llm.ts
Comment (mishig25) [inline packages/gguf/scripts/generate-llm.ts:110]:
```suggestion
		const cppConst = line.match(RE_ARCH_NAME)?.[0];
```

File: packages/gguf/scripts/generate-llm.ts
Comment (mishig25) [inline packages/gguf/scripts/generate-llm.ts:137]:
```suggestion
		const cppConst = line.match(RE_CASE)?.[1];
```

File: packages/gguf/scripts/generate-llm.ts
Comment (mishig25) [inline packages/gguf/scripts/generate-llm.ts:143]:
```suggestion
		const keyConst = line.match(/LLM_KV_[A-Z0-9_]+/)?.[0];
```

File: packages/gguf/scripts/generate-llm.ts
Comment (mishig25) [inline packages/gguf/scripts/generate-llm.ts:169]:
```suggestion
			if (a.hparams.length) {
```

File: packages/gguf/scripts/generate-llm.ts
Comment (ngxson) [reply packages/gguf/scripts/generate-llm.ts]: Good point! There should be no more `||` or `??` in my code: https://github.com/huggingface/huggingface.js/pull/640/commits/e9f076626ed532199b2ed509f15128816747218a

File: packages/gguf/src/types.ts
Comment (mishig25) [inline packages/gguf/src/types.ts:3]:
```suggestion
```

this import is not needed

File: packages/gguf/scripts/generate-llm.ts
Comment (mishig25) [reply packages/gguf/scripts/generate-llm.ts]: handled in 6962b5143a70631f5a2736b53ba1ddbdc8d609f0

File: packages/gguf/scripts/generate-llm.ts
Comment (mishig25) [reply packages/gguf/scripts/generate-llm.ts]: handled in 6962b5143a70631f5a2736b53ba1ddbdc8d609f0

File: packages/gguf/scripts/generate-llm.ts
Comment (mishig25) [reply packages/gguf/scripts/generate-llm.ts]: handled in 6962b5143a70631f5a2736b53ba1ddbdc8d609f0

File: packages/gguf/scripts/generate-llm.ts
Comment (mishig25) [reply packages/gguf/scripts/generate-llm.ts]: handled in 6962b5143a70631f5a2736b53ba1ddbdc8d609f0

File: packages/gguf/scripts/generate-llm.ts
Comment (mishig25) [reply packages/gguf/scripts/generate-llm.ts]: handled in 6962b5143a70631f5a2736b53ba1ddbdc8d609f0

File: packages/gguf/scripts/generate-llm.ts
Comment (mishig25) [reply packages/gguf/scripts/generate-llm.ts]: handled in 6962b5143a70631f5a2736b53ba1ddbdc8d609f0

File: packages/gguf/scripts/generate-llm.ts
Comment (mishig25) [reply packages/gguf/scripts/generate-llm.ts]: handled in https://github.com/huggingface/huggingface.js/commit/6962b5143a70631f5a2736b53ba1ddbdc8d609f0

File: packages/gguf/scripts/generate-llm.ts
Comment (mishig25) [reply packages/gguf/scripts/generate-llm.ts]: handled in https://github.com/huggingface/huggingface.js/commit/6962b5143a70631f5a2736b53ba1ddbdc8d609f0

File: packages/gguf/scripts/generate-llm.ts
Comment (mishig25) [reply packages/gguf/scripts/generate-llm.ts]: handled in e9f076626ed532199b2ed509f15128816747218a

File: packages/gguf/scripts/generate-llm.ts
Comment (mishig25) [reply packages/gguf/scripts/generate-llm.ts]: handled in 6962b5143a70631f5a2736b53ba1ddbdc8d609f0

File: packages/gguf/scripts/generate-llm.ts
Comment (mishig25) [reply packages/gguf/scripts/generate-llm.ts]: `([^;]+)` is fine as it is

**Comment by mishig25** [comment]
[the proposal](https://github.com/huggingface/huggingface.js/pull/640#issuecomment-2083666339) sounds good to me. 

> but I'm not sure if my proposal is the right way to do

You can start with a draft PR and we can iterate if needed

**Review by mishig25 — APPROVED** [review]
If you can write a small section on [readme](https://github.com/ngxson/huggingface.js/blob/e9f076626ed532199b2ed509f15128816747218a/packages/gguf/README.md) about this change, that would be great.

I've left one last nit here https://github.com/huggingface/huggingface.js/pull/640#discussion_r1584348110

LGTM! Unless @julien-c comments in the next few hours, I will merge this PR soon

File: packages/gguf/src/types.ts
Comment (ngxson) [reply packages/gguf/src/types.ts]: Removed with https://github.com/huggingface/huggingface.js/pull/640/commits/09b70cda0654d48005b783bae92369f81a11783c

**Comment by ngxson** [comment]
Thanks for the careful review. For now it's not very clear how to use these types, so I don't think I can add anything to readme right now. Maybe with the next PR (the proposal that you mentioned), I will add an example to readme.

File: packages/gguf/src/types.ts
Comment (julien-c) [inline packages/gguf/src/types.ts:50]:
```suggestion
const ARCHITECTURES = [...LLM_ARCHITECTURES, "rwkv", "whisper"] as const;
```

I think you need to add this here otherwise the type inference is `string[]` – unless we're not using the same version of typescript?

**Review by julien-c — APPROVED** [review]
re. your proposal, please note that I am not super good with TS type inference, but IMO the ideal way we would use this is to enable like this:

```ts
if (metadata["general.architecture"] === "grok") {
	const x = metadata; // type = TransformerLLMBase<"grok">
	const foo = x["grok.attention.alibi_bias_max"]; // type = number
}
```

do you think it's possible at all?

File: packages/gguf/src/types.ts
Comment (coyotte508) [inline packages/gguf/src/types.ts:64]: @mishig25 is this on purpose? 

`TArchitecture` will always extend `"encoder.whisper" | "decoder.whisper"`

File: packages/gguf/src/types.ts
Comment (coyotte508) [inline packages/gguf/src/types.ts:67]: It's maybe more readable to have:

```suggestion
Record<
	| `${TArchitecture}.layer_count`
	| `${TArchitecture}.feed_forward_length`
	| `${TArchitecture}.context_length`
	| `${TArchitecture}.embedding_length`
	| `${TArchitecture}.block_count`,
	number
>
```

**Comment by coyotte508** [comment]
> image
> 
> checked it manually. As of right now, ts server is just giving generic `MetadatValue` rather than specific type (e.g. `number`)
> 
> cc: @coyotte508 @ngxson do you know if above request is possible [#640 (comment)](https://github.com/huggingface/huggingface.js/pull/640#issuecomment-2085961578) ?

You can do this:

```ts
const x: TransformerLLM = {} as any;

if ("grok.layer_count" in x) {
	// below is typed correctly
	console.log(x["grok.attention.layer_norm_epsilon"])
}
```



> maybe by doing some generics stuff in
> 
> ```ts
> export type GGUFMetadata<K extends Architecture = Architecture> = {
> 	version: Version;
> 	tensor_count: bigint;
> 	kv_count: bigint;
> } & Partial<General<K>> &
> 	Partial<ModelBase<K>>;
> ```
> 
> maybe too complex though

Yes it's possible probably

You would need a mapped type:

```ts
type Archs = {
  "grok": ArchGrok,
  ...
} satisfies Record<Architecture, unknown>
```

And then call `Archs[T]` somewhere in you `Model<T>` or `GGufMetaData<T>` type.

Since we have code generation it's also possible to generate every possible `GGufMetadata` and do `GGufMetadata = GgufMetadaGrok | ...`. And have less complex generics code (but more verbose)

**Review by coyotte508 — APPROVED** [review]
LGTM, although would appreciate https://github.com/huggingface/huggingface.js/pull/640#discussion_r1585321087 :)

File: packages/gguf/src/types.ts
Comment (ngxson) [reply packages/gguf/src/types.ts:67]: Actually it's being formatted by `pnpm run format`. Can we maybe ignore specific lines from the formatter?

File: packages/gguf/src/types.ts
Comment (ngxson) [reply packages/gguf/src/types.ts:64]: I couldn't find any whisper gguf model on the internet, so maybe we should temporary remove support for whisper and we'll see later?

More info: there're some [ggml models uploaded by ggerganov](https://huggingface.co/ggerganov/whisper.cpp/tree/main), but they uses ggml format, not gguf.

File: packages/gguf/src/types.ts
Comment (julien-c) [reply packages/gguf/src/types.ts:67]: maybe we merge this one and improve in later PRs?

Comment (ngxson) [reply packages/gguf/src/types.ts:67]: Yeah I think we can merge now. I'm not having much time for doing code atm, but I'll open another pull to resolve this and the also the proposal related to usage of these types (hopefully I'll have time this weekend).

File: packages/gguf/src/types.ts
Comment (mishig25) [reply packages/gguf/src/types.ts:64]: the whisper spec of gguf is here https://github.com/ggerganov/ggml/blob/master/docs/gguf.md#whisper

Comment (mishig25) [reply packages/gguf/src/types.ts:64]: > @mishig25 is this on purpose?

yes

**Comment by mishig25** [comment]
Let's merge ?

**Comment by mishig25** [comment]
@ngxson thanks a lot for the contribution ! 
looking forward to your future contrib 🔥 https://github.com/huggingface/huggingface.js/pull/640#discussion_r1587394771

**Pull request merged**
